### ISSUE TITLE
how to get more game-specific key of other games for LAN protocol

### ISSUE THREAD

**Issue by ticty** [kinnay/NintendoClients#58]
Hello :-)
How can I get more game-specific key of other games for LAN protocol ? eg. key for Mario Kart 8 Deluxe.
Is there a way that I can do it myself ?

**Comment by kinnay** [comment]
Hi,

> Is there a way that I can do it myself ?

You have to disassemble and reverse engineer the game. There is no other way.

I just tried to figure out the key myself, but I'm not sure if it has a game-specific key at all. I know that Mario Kart 8 uses the ENL library, so I thought that it uses [the same key generation algorithm](https://github.com/kinnay/NintendoClients/wiki/ENL-Key-Generation) as Splatoon 2 and SMM 2, but Mario Kart 8 seems to use an older version of the library that doesn't have the key generation algorithm yet.

The Pia version of MK8D is 5.6, which I haven't seen before. I know that the crypto challenge was added to the LAN protocol somewhere between version 5.4 and 5.9. This makes me believe that 5.6 does not have the crypto challenge yet, which would mean that it doesn't need a game specific key. I wonder how it gets the session key for actual game communication in that case though. Maybe the traffic isn't encrypted at all.

Anyway, for Mario Kart 8 Deluxe, this needs more research :slightly_smiling_face:

**Comment by ticty** [comment]
Thank you for your reply!
I will try to analyze the PIA packets of MK8D compares to splatoon2.

> You have to disassemble and reverse engineer the game. There is no other way

 So can I download the game files somewhere from internet and disassemble it without an unlocked swtich?

**Comment by kinnay** [comment]
> So can I download the game files somewhere from internet and disassemble it without having an unlocked switch?

Maybe. It's better to dump the files from your switch though. But that's only possible if your switch is hackable of course.

**Comment by kinnay** [comment]
Ok Mario Kart 8 indeed has no crypto challenge in the LAN protocol. I updated the wiki page about the LAN protocol. I also fixed the code in the python package.

If you only want to search for LAN sessions the following code now works fine for MK8:

```python
from nintendo.pia import lan, settings
import anyio

async def main():
	s = settings.default(50600)
	crit = lan.LanSessionSearchCriteria()
	sessions = await lan.browse(s, crit)
	print(sessions)

anyio.run(main)
```
I don't know how the traffic outside of the LAN protocol is encrypted yet. In Pia 5.7 and later the session key is derived from the 'session param', but since MK8 uses 5.6 there are no session param in the browse reply.
